### PR TITLE
Enable HTTP/3 for HttpSys (and IIS?) 

### LINKED ISSUE #29077: Http/3 for HttpSys
We'll need some basic tests for Http/3 support on HttpSys including:
- The bindings work as expected
- The protocol version is correctly reported
- Hello World requests
- gRPC scenarios (Reset, Trailers)
- Interop with HttpClient

We can use the new windows preview helix queue to run these.

### PR DESCRIPTION
**Pull request by Tratcher**
Fixes #29077. Http protocols are entirely abstracted away by Http.Sys so the only real change here is to check a new bitflag to get the right version. They don't update the version field for back compat reasons.

In theory these changes will enable IIS to work as well, but that's not tested here.

### CODE REVIEW

File: src/Servers/HttpSys/test/FunctionalTests/MsQuicSupportedAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Net.Quic;
+using Microsoft.AspNetCore.Testing;
+
+namespace Microsoft.AspNetCore.Server.HttpSys
```
Comment (wtgodbe) [inline src/Servers/HttpSys/test/FunctionalTests/MsQuicSupportedAttribute.cs:8]: Can't share this w/ the one in Kestrel?

**Review by wtgodbe — APPROVED** [review]

File: src/Servers/HttpSys/test/FunctionalTests/Http3SupportedAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+using System;
+using System.Net.Quic;
+using Microsoft.AspNetCore.Testing;
+using Microsoft.Win32;
+
+namespace Microsoft.AspNetCore.Server.HttpSys
+{
+    [AttributeUsage(AttributeTargets.Method | AttributeTargets.Class, AllowMultiple = false)]
+    public class Http3SupportedAttribute : Attribute, ITestCondition
```
Comment (JamesNK) [inline src/Servers/HttpSys/test/FunctionalTests/Http3SupportedAttribute.cs:12]: I don't see anything that uses this attribute.

File: src/Servers/HttpSys/test/FunctionalTests/Http3Tests.cs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+// Not tested here: Http.Sys supports sending an altsvc HTTP/2 frame if you enable the following reg key.
+// reg add "HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\services\HTTP\Parameters" /v EnableAltSvc /t REG_DWORD /d 1 /f
+// However, this only works with certificate bindings that specify a name. We test with the IP based bindings created by IIS Express.
+// I don't know if the client supports the HTTP/2 altsvc frame.
+namespace Microsoft.AspNetCore.Server.HttpSys
+{
+    [MsQuicSupported] // Required by HttpClient
+    [Http3Supported]
+    public class Http3Tests
```
Comment (JamesNK) [inline src/Servers/HttpSys/test/FunctionalTests/Http3Tests.cs:18]: Is there any need to test features like reset and trailing headers still work with HTTP/3?

File: src/Servers/HttpSys/test/FunctionalTests/Http3SupportedAttribute.cs (hunk 2 above)
Comment (JamesNK) [reply src/Servers/HttpSys/test/FunctionalTests/Http3SupportedAttribute.cs:12]: Never mind, I searched for Http3SupportedAttribute but there is `[Http3Supported]`.

File: src/Servers/HttpSys/test/FunctionalTests/Http3SupportedAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+        public string SkipReason => "HTTP/3 is not supported or enabled on the current test machine";
+
+        private static bool IsRegKeySet
+        {
+            get
+            {
+                try
+                {
+                    using var key = Registry.LocalMachine.OpenSubKey(@"SYSTEM\CurrentControlSet\services\HTTP\Parameters");
```
Comment (JamesNK) [inline src/Servers/HttpSys/test/FunctionalTests/Http3SupportedAttribute.cs:25]: Add a comment that this regkey is required for HTTP/3 with HttpSys.

File: src/Servers/HttpSys/test/FunctionalTests/Http3Tests.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                    Assert.True(httpContext.Request.IsHttps);
+                    await httpContext.Response.WriteAsync(httpContext.Request.Protocol);
+                }
+                catch (Exception ex)
+                {
+                    await httpContext.Response.WriteAsync(ex.ToString());
+                }
+            });
+            var handler = new HttpClientHandler();
+            // Needed on CI, the IIS Express cert we ues isn't trusted there.
```
Comment (adityamandaleeka) [inline src/Servers/HttpSys/test/FunctionalTests/Http3Tests.cs:36]:
```suggestion
            // Needed on CI, the IIS Express cert we use isn't trusted there.
```

File: src/Servers/HttpSys/test/FunctionalTests/Http3Tests.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                }
+                catch (Exception ex)
+                {
+                    await httpContext.Response.WriteAsync(ex.ToString());
+                }
+            });
+
+            altsvc = $@"h3="":{new Uri(address).Port}""";
+            var handler = new HttpClientHandler();
+            // Needed on CI, the IIS Express cert we ues isn't trusted there.
```
Comment (adityamandaleeka) [inline src/Servers/HttpSys/test/FunctionalTests/Http3Tests.cs:66]:
```suggestion
            // Needed on CI, the IIS Express cert we use isn't trusted there.
```

File: src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs [hunk 7]
```diff
@@ -192,19 +195,23 @@ internal CookedUrl GetCookedUrl()
 
         internal Version GetVersion()
         {
+            if (IsHttp3)
+            {
+                return HttpVersion.Version30;
+            }
             if (IsHttp2)
             {
-                return Constants.V2;
```
Comment (adityamandaleeka) [inline src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs:197]: Not for this PR but any reason why we can't clean up these versions from Constants?

**Review by adityamandaleeka — APPROVED** [review]

File: src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs (hunk 7 above)
Comment (Tratcher) [reply src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs:197]: They're used in several other places that I didn't want to mess with at the moment.

Comment (adityamandaleeka) [reply src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs:197]: Yup, like I said, not right now. Sounds like it's a good beginner-friendly cleanup issue.

File: src/Servers/HttpSys/test/FunctionalTests/Http3Tests.cs (hunk 3 above)
Comment (Tratcher) [reply src/Servers/HttpSys/test/FunctionalTests/Http3Tests.cs:20]: Added. I ended up adding one special case in the code for abort codes.

**Review by JamesNK — APPROVED** [review]

**Pull request merged**